src/engine/TransitivePath.cpp

```diff
@@ -12,31 +12,29 @@
 #include "util/Exception.h"
 
 // _____________________________________________________________________________
-TransitivePath::TransitivePath(
-    QueryExecutionContext* qec, std::shared_ptr<QueryExecutionTree> child,
-    bool leftIsVar, bool rightIsVar, size_t leftSubCol, size_t rightSubCol,
-    Id leftValue, Id rightValue, const Variable& leftColName,
-    const Variable& rightColName, size_t minDist, size_t maxDist)
+TransitivePath::TransitivePath(QueryExecutionContext* qec,
+                               std::shared_ptr<QueryExecutionTree> child,
+                               TransitivePathSide leftSide,
+                               TransitivePathSide rightSide, size_t minDist,
+                               size_t maxDist)
     : Operation(qec),
-      _leftSideTree(nullptr),
-      _leftSideCol(-1),
-      _rightSideTree(nullptr),
-      _rightSideCol(-1),
-      _resultWidth(2),
       _subtree(std::move(child)),
-      _leftIsVar(leftIsVar),
-      _rightIsVar(rightIsVar),
-      _leftSubCol(leftSubCol),
-      _rightSubCol(rightSubCol),
-      _leftValue(leftValue),
-      _rightValue(rightValue),
-      // TODO<joka921> Those should also be variables.
-      _leftColName(leftColName.name()),
-      _rightColName(rightColName.name()),
+      _lhs(leftSide),
+      _rhs(rightSide),
+      _resultWidth(2),
       _minDist(minDist),
       _maxDist(maxDist) {
-  _variableColumns[leftColName] = makeAlwaysDefinedColumn(0);
-  _variableColumns[rightColName] = makeAlwaysDefinedColumn(1);
+  if (leftSide.isVariable()) {
+    _variableColumns[std::get<Variable>(leftSide.value)] =
+        makeAlwaysDefinedColumn(0);
+  }
+  if (rightSide.isVariable()) {
+    _variableColumns[std::get<Variable>(rightSide.value)] =
+        makeAlwaysDefinedColumn(1);
+  }
```

Comment:
Could you clarify what you mean by valid? Which `TransitivePathSide`s would not be valid?

Reply:
Should the class support passing in a bound side via the constructor instead of `bindLeftOrRightSide`? Or should it throw an exception? From my understanding, there is currently no place in the code where a side is bound without using the `bindLeftOrRightSide` function.

I think you make a good point about using the variant. I'm pretty sure there are more places where this would simplify the code. However, this means a lot more changes and I think it would be better to do this in another PR with the other more involved changes you proposed.